### PR TITLE
add BIN for string function support

### PR DESCRIPTION
**Pull request by a6802739**
I have add a function `BIN` for #310.

cc @siddontang @coocood.

I wonder If I should take `negative number` into consider, I'm not sure if `mysql` support `negative number` now?

### CODE REVIEW

**Comment by a6802739** [comment]
@siddontang , How could I rebase the two commit into one now.

**Comment by shenli** [comment]
@a6802739 Thanks for your PR!
You can use git squash command or We can squash your commits when merging your PR.

**Comment by shenli** [comment]
@a6802739 You also need to add bin function syntax in parser package. I will send you a document to describe how to do this.

**Comment by a6802739** [comment]
@shenli,  Thanks for you review. So would you mind send the document to my email `895179504@qq.com`.

File: evaluator/builtin_string.go [hunk 1]
```diff
@@ -50,6 +50,60 @@ func builtinLength(args []types.Datum, _ context.Context) (d types.Datum, err er
...
+	var bs string
+	switch args[0].Kind() {
+	case types.KindNull:
+		return d, nil
+	default:
+		v, err := args[0].ToInt64(ctx.GetSessionVars().StmtCtx)
+		if err != nil {
+			return d, errors.Trace(err)
+		}
+		switch {
```
Comment (hanfei1991) [inline evaluator/builtin_string.go]: You don't need to convert an integer to bytes, just use bit operator :)
> 
while(v > 0){
    if (v & 1 > 0){
         bs = bs + '1'
    } else {
        bs = bs + '0'
    }
    v = v >> 1
}

Comment (a6802739) [inline evaluator/builtin_string.go]: oh, got it.

Comment (a6802739) [inline evaluator/builtin_string.go]: we should consider it's `bigendian` or `littleendian`.

Comment (hanfei1991) [inline evaluator/builtin_string.go]: BE and LE is about network transporting, needn't be considered here.

Comment (a6802739) [inline evaluator/builtin_string.go]: for(v > 0){
if (v & 1 > 0){
bs = append(bs, "1")
} else {
bs = append(bs, "0")
}
v = v >> 1
}
i := 0
length := len(bs)
for i < length / 2 {
   temp := bs[i]
   bs[i] = bs[length-i-1]
   bs[length-i-1] = temp
   i = i + 1
}
rs := strings.Join(bs, "")

Comment (coocood) [inline evaluator/builtin_string.go]: Appending character '0' and '1' to a byte slice is more efficient than append strings to a string slice.

**Comment by coocood** [comment]
@a6802739 
I think we better use standard library to implement this function.
```
fmt.Sprintf("%b", n)
```
or
```
strconv.FormatInt(n, 2)
```

**Comment by coocood** [comment]
@a6802739 
I tried to run `select bin(-1)` on MySQL, the result is `1111111111111111111111111111111111111111111111111111111111111111`

I seem like a negative integer will be converted to unsigned interger first.

**Comment by a6802739** [comment]
@coocood, `fmt.Sprintf("%b", n)` is good.

`1111111111111111111111111111111111111111111111111111111111111111` is `^(-1) + 1`, right?

So, it seems `mysql` support negative integer.

**Comment by coocood** [comment]
@a6802739 
```
x := int64(-1)
fmt.Printf("%b", uint64(x))
```

**Comment by a6802739** [comment]
okay. got it.

**Comment by coocood** [comment]
@a6802739  
You can search keyword `builtin` in PR list, find a PR that adds a new builtin function, see what is needed to add a new function.

**Comment by coocood** [comment]
@a6802739 
Please make sure test is passed before push a commit.

**Comment by ngaut** [comment]
PTAL

**Comment by a6802739** [comment]
@coocood, where is `scanner.l`?

**Comment by coocood** [comment]
@a6802739 
I'm sorry that PR is outdated, checkout this PR instead:
https://github.com/pingcap/tidb/pull/2249

**Comment by shenli** [comment]
@a6802739 Any update? 
We write an article to show how to add builtin function: https://github.com/ngaut/builddatabase/blob/master/tidb/builtin.md

**Comment by coocood** [comment]
@a6802739 
It has been a long time since the last update.
Close for now, if you have any update, I'll reopen it.

**Pull request closed without merging**